src/extension/kernel.ts

```diff
@@ -237,14 +248,170 @@ export class Kernel implements Disposable {
     exec.start(Date.now())
     let execKey = getKey(runningCell)
 
-    /**
-     * check if user is running experiment to execute shell via runme cli
-     */
     const hasPsuedoTerminalExperimentEnabled = this.hasExperimentEnabled('pseudoterminal')
     const terminal = this.#terminals.get(cell.document.uri.fsPath)
-    const successfulCellExecution = (hasPsuedoTerminalExperimentEnabled && terminal)
-      ? await runme.call(this, exec, terminal)
-      : await executor[execKey].call(this, exec, runningCell)
+
+    let successfulCellExecution: boolean
+
+    // TODO: handle connection failures better!!
+    if(
+      this.runner && 
+      !(hasPsuedoTerminalExperimentEnabled && terminal) &&
+      (execKey === 'bash' || execKey === 'sh')
+    ) {
+      // TODO: bring these to top level in new file
+      const RUNME_ID = `${runningCell.fileName}:${exec.cell.index}`
+      const LABEL_LIMIT = 15
+
+      const cwd = path.dirname(runningCell.uri.fsPath)
+      
+      const execute = async (resolve: (res: boolean) => void) => {
+        const runner = this.runner!
+        const cellText = exec.cell.document.getText()
+        const script = getCmdShellSeq(cellText, PLATFORM_OS)
+
+        const annotations = getAnnotations(exec.cell)
+        const { interactive, mimeType, background } = annotations
+        
+        const program = await runner.createProgramSession({
+          // TODO: make this work on windows/if bash isn't installed/etc
+          programName: execKey,
+          script: {
+            type: 'script', script
+          },
+          envs: [
+            `RUNME_ID=${RUNME_ID}`
+          ],
+          cwd,
+          tty: interactive
+        })
+
+        program.onDidClose((code) => {
+          if(code) {
+            // TODO: do something with this code
+            resolve(false)
+          }
+
+          resolve(true)
+        })
+
+        if(!interactive) {
+          const MIME_TYPES_WITH_CUSTOM_RENDERERS = ['text/plain']
```

Comment:
I would suggest to move this to constants file

Reply:
Before merge this whole function will be brought out, probably into its own file. Constants will be put at top-level as part of that process 👍